packages/astro/src/actions/integration.ts

```diff
@@ -0,0 +1,52 @@
+import type { AstroIntegration, AstroSettings } from '../@types/astro.js';
+import { ActionsWithoutServerOutputError } from '../core/errors/errors-data.js';
+import { AstroError } from '../core/errors/errors.js';
+import { isServerLikeOutput, viteID } from '../core/util.js';
+import { ACTIONS_TYPES_FILE, VIRTUAL_MODULE_ID } from './consts.js';
+
+/**
+ * This integration is applied when the user is using Actions in their project.
+ * It will inject the necessary routes and middlewares to handle actions.
+ */
+export default function astroIntegrationActionsRouteHandler({
+	settings,
+}: {
+	settings: AstroSettings;
+}): AstroIntegration {
+	return {
+		name: VIRTUAL_MODULE_ID,
+		hooks: {
+			async 'astro:config:setup'(params) {
+				if (!isServerLikeOutput(params.config)) {
```

Comment:
I think this should occur in config:done instead since it could be updated during any integration config:setup

Reply:
Oh good point. I guess it would error _after_ we attempt to add middleware, but I guess that's fine